File: RoutinUsage/Providers/CommandCodeUsageProvider.swift
import Foundation
import CryptoKit

struct CommandCodeUsageProvider: UsageProvider {
    static let baseURL = URL(string: "https://api.commandcode.ai")!

    let descriptor: ProviderDescriptor
    private let session: URLSession
    private let metadataCache: CommandCodeMetadataCache

    init(session: URLSession = .shared) {
        self.session = session
        self.metadataCache = CommandCodeMetadataCache()
        self.descriptor = ProviderRegistry.builtInDescriptors.first(where: { $0.id == .commandCode })!
    }

    func metricCapabilities(for configuration: KeyConfiguration) -> [UsageMetricCapability] {
        guard configuration.credentialKind == .bearerAPIKey else { return [] }
        return [
            UsageMetricCapability(
                metricID: "credit-progress",
                label: "月",
                presentation: .progress,
                semantic: .usedQuota,
                isMenuBarSelectable: true,
                menuBarPriority: 0,
                defaultAlertEnabled: true,
                defaultAbsoluteAlertThreshold: nil
            ),
            UsageMetricCapability(
                metricID: "five-hour",
                label: "5 小时",
                presentation: .progress,
                semantic: .usedQuota,
                isMenuBarSelectable: true,
                menuBarPriority: 1,
                defaultAlertEnabled: true,
                defaultAbsoluteAlertThreshold: nil
            ),
            UsageMetricCapability(
                metricID: "weekly",
                label: "周",
                presentation: .progress,
                semantic: .usedQuota,
                isMenuBarSelectable: true,
                menuBarPriority: 2,
                defaultAlertEnabled: true,
                defaultAbsoluteAlertThreshold: nil
            ),
            valueCapability(metricID: "credit-balance", label: "剩余额度"),
            valueCapability(metricID: "monthly-remaining", label: "月度剩余"),
            valueCapability(metricID: "purchased-remaining", label: "购买剩余"),
            valueCapability(metricID: "free-remaining", label: "赠送剩余"),
            valueCapability(metricID: "period-spent", label: "本周期消费"),
            valueCapability(metricID: "request-count", label: "累计请求")
        ]
    }

    func validate(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        try await fetchUsage(credential, now: now)
    }

    func fetchUsage(_ credential: ProviderCredential, now: Date) async throws -> UsageSnapshot? {
        guard credential.providerID == .commandCode,
              credential.kind == .bearerAPIKey,
              !credential.secret.trimmingCharacters(in: .whitespacesAndNewlines).isEmpty
        else { throw UsageProviderError.invalidCredential }

        let cacheKey = Self.cacheKey(for: credential)
        let cached = await metadataCache.snapshot(for: cacheKey)

        let whoamiTask = Task<CommandCodeWhoami, Error> {
            try await request(
                "/alpha/whoami",
                query: [URLQueryItem(name: "limits", value: "1")],
                credential: credential
            )
        }

        let creditsTask = Task {
            do {
                return Result<CommandCodeCreditsResponse, Error>.success(
                    try await request(
                        "/alpha/billing/credits",
                        query: [URLQueryItem(name: "orgId", value: cached.orgID)],
                        credential: credential
                    )
                )
            } catch {
                return .failure(error)
            }
        }

        let subscriptionsTask = Task {
            do {
                return Result<CommandCodeSubscriptionResponse, Error>.success(
                    try await request(
                        "/alpha/billing/subscriptions",
                        query: [URLQueryItem(name: "orgId", value: cached.orgID)],
                        credential: credential
                    )
                )
            } catch {
                return .failure(error)
            }
        }

        let modelsTask = Task<[String], Never> {
            await fetchModelIDs(
                cacheKey: cacheKey,
                credential: credential,
                now: now
            )
        }

        let preloadedSummaryTask = cached.periodStart.map { periodStart in
            Task<Result<CommandCodeUsageSummary, Error>, Never> {
                do {
                    return .success(
                        try await request(
                            "/alpha/usage/summary",
                            query: [
                                URLQueryItem(name: "orgId", value: cached.orgID),
                                URLQueryItem(name: "since", value: periodStart)
                            ],
                            credential: credential
                        )
                    )
                } catch {
                    return .failure(error)
                }
            }
        }

        let whoami: CommandCodeWhoami
        do {
            whoami = try await whoamiTask.value
        } catch let error as UsageProviderError {
            creditsTask.cancel()
            subscriptionsTask.cancel()
            modelsTask.cancel()
            throw error
        } catch {
            creditsTask.cancel()
            subscriptionsTask.cancel()
            modelsTask.cancel()
            throw UsageProviderError.invalidResponse
        }
        let actualOrgID = whoami.org?.id

        let creditsResponse: CommandCodeCreditsResponse
        let subscriptionResponse: CommandCodeSubscriptionResponse
        do {
            if actualOrgID == cached.orgID {
                creditsResponse = try await creditsTask.value.get()
                subscriptionResponse = try await subscriptionsTask.value.get()
            } else {
                creditsTask.cancel()
                subscriptionsTask.cancel()
                creditsResponse = try await request(
                    "/alpha/billing/credits",
                    query: [URLQueryItem(name: "orgId", value: actualOrgID)],
                    credential: credential
                )
                subscriptionResponse = try await request(
                    "/alpha/billing/subscriptions",
                    query: [URLQueryItem(name: "orgId", value: actualOrgID)],
                    credential: credential
                )
            }
        } catch let error as UsageProviderError {
            throw error
        } catch {
            throw UsageProviderError.invalidResponse
        }

        guard let credits = creditsResponse.credits else {
            throw UsageProviderError.providerMessage("Command Code：额度信息返回失败")
        }
        let subscription = subscriptionResponse.data
        let currentPeriodStart = subscription?.currentPeriodStart
        let summary: CommandCodeUsageSummary
        do {
            if let summaryTask = preloadedSummaryTask,
               actualOrgID == cached.orgID,
               currentPeriodStart == cached.periodStart {
                summary = try await summaryTask.value.get()
            } else {
                preloadedSummaryTask?.cancel()
                summary = try await request(
                    "/alpha/usage/summary",
                    query: [
                        URLQueryItem(name: "orgId", value: actualOrgID),
                        URLQueryItem(name: "since", value: currentPeriodStart)
                    ],
                    credential: credential
                )
            }
        } catch let error as UsageProviderError {
            throw error
        } catch {
            throw UsageProviderError.invalidResponse
        }

        await metadataCache.update(
            for: cacheKey,
            orgID: actualOrgID,
            periodStart: currentPeriodStart
        )

        let monthlyRemaining = max(0, credits.monthlyCredits ?? 0)
        let allowedModels = await modelsTask.value
        let purchasedRemaining = max(0, credits.purchasedCredits ?? 0)
        let freeRemaining = max(0, credits.freeCredits ?? 0)
        let totalRemaining = monthlyRemaining + purchasedRemaining + freeRemaining
        let totalSpent = max(0, summary.totalCost ?? 0)
        let plan = CommandCodePlanInfo(planID: subscription?.planID)
        let isActivePlan = subscription?.status == "active" || subscription?.status == "trialing"
        let totalPool: Decimal
        if isActivePlan, let monthlyAllowance = plan.monthlyCredits {
            totalPool = max(monthlyAllowance, monthlyRemaining) + purchasedRemaining + freeRemaining
        } else {
            totalPool = totalSpent + totalRemaining
        }
        let totalUsed = max(0, totalPool - totalRemaining)
        let health = UsageMetricHealthEvaluator.balanceState(
            balance: totalRemaining,
            warningThreshold: nil,
            isAvailable: true
        )

        var metrics = [
            progressMetric(
                id: "credit-progress",
                label: "月",
                used: totalUsed,
                limit: totalPool,
                remaining: totalRemaining,
                healthState: health
            ),
            valueMetric(id: "credit-balance", label: "剩余额度", value: totalRemaining, health: health),
            valueMetric(id: "monthly-remaining", label: "月度剩余", value: monthlyRemaining, health: health),
            valueMetric(id: "purchased-remaining", label: "购买剩余", value: purchasedRemaining, health: health),
            valueMetric(id: "free-remaining", label: "赠送剩余", value: freeRemaining, health: health),
            valueMetric(id: "period-spent", label: "本周期消费", value: totalSpent, health: .normal),
            valueMetric(
                id: "request-count",
                label: "累计请求",
                value: Decimal(max(0, summary.totalCount ?? 0)),
                health: .normal
            )
        ]
        metrics.append(contentsOf: windowMetrics(from: creditsResponse.windowLimits))

        return UsageSnapshot(
            planName: plan.name,
            planId: subscription?.planID,
            kind: .periodic,
            fiveHour: nil,
            weekly: nil,
            token: nil,
            allowedModels: allowedModels,
            fetchedAt: now,
            statusText: CommandCodePlanInfo.statusText(subscription?.status),
            billingMode: subscription?.status,
            subscriptionStartAt: Self.date(currentPeriodStart),
            subscriptionEndAt: Self.date(subscription?.currentPeriodEnd),
            providerID: .commandCode,
            credentialID: credential.credentialID,
            metrics: metrics
        )
    }

    private func fetchModelIDs(
        cacheKey: String,
        credential: ProviderCredential,
        now: Date
    ) async -> [String] {
        let cached = await metadataCache.snapshot(for: cacheKey)
        if let fetchedAt = cached.modelsFetchedAt,
           now.timeIntervalSince(fetchedAt) < Self.modelsCacheLifetime {
            return cached.models
        }

        do {
            let response: CommandCodeModelsResponse = try await request(
                "/provider/v1/models",
                query: [],
                credential: credential,
                timeoutInterval: 5
            )
            let models = response.data.compactMap {
                let id = $0.id.trimmingCharacters(in: .whitespacesAndNewlines)
                return id.isEmpty ? nil : id
            }
            await metadataCache.updateModels(
                for: cacheKey,
                models: models,
                fetchedAt: now
            )
            return models
        } catch {
            return cached.models
        }
    }

    private static func cacheKey(for credential: ProviderCredential) -> String {
        let digest = SHA256.hash(data: Data(credential.secret.utf8))
        let secretDigest = digest.map { String(format: "%02x", $0) }.joined()
        if let credentialID = credential.credentialID {
            return "\(credentialID.uuidString):\(secretDigest)"
        }
        return secretDigest
    }

    private static let modelsCacheLifetime: TimeInterval = 24 * 60 * 60

    private func windowMetrics(
        from limits: CommandCodeWindowLimits?
    ) -> [NormalizedUsageMetric] {
        guard limits?.limited == true else { return [] }
        var metrics: [NormalizedUsageMetric] = []
        if let fiveHour = limits?.fiveHour {
            metrics.append(windowMetric(id: "five-hour", label: "5 小时", limit: fiveHour))
        }
        if let weekly = limits?.weekly {
            metrics.append(windowMetric(id: "weekly", label: "周", limit: weekly))
        }
        return metrics
    }

    private func windowMetric(
        id: String,
        label: String,
        limit: CommandCodeUsageWindow
    ) -> NormalizedUsageMetric {
        let used = max(0, limit.used ?? 0)
        let cap = max(0, limit.cap ?? 0)
        return NormalizedUsageMetric(
            id: id,
            label: label,
            used: used,
            limit: cap,
            remaining: max(0, cap - used),
            unit: .currency,
            windowEnd: Self.date(fromMilliseconds: limit.resetAt),
            presentation: .progress,
            semantic: .usedQuota,
            currencyCode: "$",
            healthState: .normal
        )
    }

    private func progressMetric(
        id: String,
        label: String,
        used: Decimal,
        limit: Decimal,
        remaining: Decimal,
        healthState: UsageMetricHealthState
    ) -> NormalizedUsageMetric {
        NormalizedUsageMetric(
            id: id,
            label: label,
            used: used,
            limit: limit,
            remaining: remaining,
            unit: .currency,
            presentation: .progress,
            semantic: .usedQuota,
            currencyCode: "$",
            healthState: healthState
        )
    }

    private func valueMetric(
        id: String,
        label: String,
        value: Decimal,
        health: UsageMetricHealthState
    ) -> NormalizedUsageMetric {
        NormalizedUsageMetric(
            id: id,
            label: label,
            value: value,
            unit: .currency,
            presentation: .value,
            semantic: .value,
            currencyCode: "$",
            healthState: health
        )
    }

    private func valueCapability(metricID: String, label: String) -> UsageMetricCapability {
        UsageMetricCapability(
            metricID: metricID,
            label: label,
            presentation: .value,
            semantic: .value,
            isMenuBarSelectable: false,
            menuBarPriority: nil,
            defaultAlertEnabled: false,
            defaultAbsoluteAlertThreshold: nil
        )
    }

    private func request<T: Decodable & Sendable>(
        _ path: String,
        query: [URLQueryItem],
        credential: ProviderCredential,
        timeoutInterval: TimeInterval = 15
    ) async throws -> T {
        var components = URLComponents(url: Self.baseURL, resolvingAgainstBaseURL: false)
        components?.path = path
        components?.queryItems = query.filter { $0.value != nil }
        guard let url = components?.url else { throw UsageProviderError.invalidCredential }

        var urlRequest = URLRequest(url: url)
        urlRequest.httpMethod = "GET"
        urlRequest.timeoutInterval = timeoutInterval
        urlRequest.setValue("Bearer \(credential.secret)", forHTTPHeaderField: "Authorization")
        urlRequest.setValue("application/json", forHTTPHeaderField: "Accept")

        let data: Data
        let response: URLResponse
        do {
            (data, response) = try await session.data(for: urlRequest)
        } catch is CancellationError {
            throw CancellationError()
        } catch {
            throw UsageProviderError.transport
        }

        guard let httpResponse = response as? HTTPURLResponse else {
            throw UsageProviderError.invalidResponse
        }
        guard (200..<300).contains(httpResponse.statusCode) else {
            switch httpResponse.statusCode {
            case 401, 403:
                throw UsageProviderError.unauthorized
            case 429:
                throw UsageProviderError.rateLimited
            case 500...599:
                throw UsageProviderError.providerUnavailable
            default:
                throw UsageProviderError.invalidResponse
            }
        }

        if let errorEnvelope = try? JSONDecoder().decode(CommandCodeErrorEnvelope.self, from: data),
           errorEnvelope.success == false {
            throw UsageProviderError.providerMessage(
                "Command Code：\(errorEnvelope.error?.message ?? "请求返回失败")"
            )
        }
        do {
            return try JSONDecoder().decode(T.self, from: data)
        } catch {
            await AppLogStore.shared.log(
                level: .error,
                event: "command_code_decode_failed",
                details: "path=\(path) reason=\(error)"
            )
            throw UsageProviderError.invalidResponse
        }
    }

    private static func date(_ value: String?) -> Date? {
        guard let value, !value.isEmpty else { return nil }
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        if let date = formatter.date(from: value) { return date }
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.date(from: value)
    }

    private static func date(fromMilliseconds value: Double?) -> Date? {
        guard let value else { return nil }
        return Date(timeIntervalSince1970: value / 1_000)
    }
}

private actor CommandCodeMetadataCache {
    struct Snapshot: Sendable {
        var orgID: String? = nil
        var periodStart: String? = nil
        var models: [String] = []
        var modelsFetchedAt: Date? = nil
    }

    private var values: [String: Snapshot] = [:]

    func snapshot(for key: String) -> Snapshot {
        values[key] ?? Snapshot()
    }

    func update(for key: String, orgID: String?, periodStart: String?) {
        var value = values[key] ?? Snapshot()
        value.orgID = orgID
        value.periodStart = periodStart
        values[key] = value
    }

    func updateModels(for key: String, models: [String], fetchedAt: Date) {
        var value = values[key] ?? Snapshot()
        value.models = models
        value.modelsFetchedAt = fetchedAt
        values[key] = value
    }
}

private struct CommandCodeErrorEnvelope: Decodable, Sendable {
    let success: Bool?
    let error: CommandCodeAPIError?
}

private struct CommandCodeAPIError: Decodable, Sendable {
    let message: String?
}

private struct CommandCodeWhoami: Decodable, Sendable {
    let user: CommandCodeUser?
    let org: CommandCodeOrg?
}

private struct CommandCodeUser: Decodable, Sendable {
    let id: String?
    let userName: String?
}

private struct CommandCodeOrg: Decodable, Sendable {
    let id: String?
    let login: String?
}

private struct CommandCodeCreditsResponse: Decodable, Sendable {
    let credits: CommandCodeCredits?
    let windowLimits: CommandCodeWindowLimits?
}

private struct CommandCodeCredits: Decodable, Sendable {
    let planID: String?
    let monthlyCredits: Decimal?
    let purchasedCredits: Decimal?
    let freeCredits: Decimal?

    private enum CodingKeys: String, CodingKey {
        case planID = "planId"
        case monthlyCredits
        case purchasedCredits
        case freeCredits
    }
}

private struct CommandCodeWindowLimits: Decodable, Sendable {
    let limited: Bool?
    let fiveHour: CommandCodeUsageWindow?
    let weekly: CommandCodeUsageWindow?
}

private struct CommandCodeUsageWindow: Decodable, Sendable {
    let used: Decimal?
    let cap: Decimal?
    let resetAt: Double?
}

private struct CommandCodeSubscriptionResponse: Decodable, Sendable {
    let data: CommandCodeSubscription?
}

private struct CommandCodeSubscription: Decodable, Sendable {
    let planID: String?
    let status: String?
    let currentPeriodStart: String?
    let currentPeriodEnd: String?

    private enum CodingKeys: String, CodingKey {
        case planID = "planId"
        case status
        case currentPeriodStart
        case currentPeriodEnd
    }
}

private struct CommandCodeUsageSummary: Decodable, Sendable {
    let totalCost: Decimal?
    let totalCount: Int?
}

private struct CommandCodeModelsResponse: Decodable, Sendable {
    let data: [CommandCodeModel]
}

private struct CommandCodeModel: Decodable, Sendable {
    let id: String
}

private struct CommandCodePlanInfo {
    static func statusText(_ status: String?) -> String? {
        switch status {
        case "active": return "有效"
        case "trialing": return "试用中"
        case "past_due": return "逾期"
        case "canceled", "cancelled": return "已取消"
        case let .some(value) where !value.isEmpty: return value
        case .some, .none: return nil
        }
    }

    let name: String
    let monthlyCredits: Decimal?

    init(planID: String?) {
        switch planID {
        case "individual-go":
            name = "Go"
            monthlyCredits = 10
        case "individual-goat":
            name = "GOAT"
            monthlyCredits = 70
        case "individual-pro", "individual-pro-v1":
            name = "Pro"
            monthlyCredits = planID == "individual-pro" ? 30 : 80
        case "individual-provider":
            name = "Provider"
            monthlyCredits = 15
        case "individual-max":
            name = "Max"
            monthlyCredits = 150
        case "individual-ultra":
            name = "Ultra"
            monthlyCredits = 300
        case "teams-pro":
            name = "Teams Pro"
            monthlyCredits = 40
        case let .some(planID):
            name = planID
            monthlyCredits = nil
        case .none:
            name = ""
            monthlyCredits = nil
        }
    }
}

File: android/provider/src/main/kotlin/ai/routin/mytoken/provider/commandcode/CommandCodeUsageProvider.kt
package ai.routin.mytoken.provider.commandcode

import ai.routin.mytoken.domain.model.Credential
import ai.routin.mytoken.domain.model.CredentialKind
import ai.routin.mytoken.domain.model.CredentialSecret
import ai.routin.mytoken.domain.model.ProviderId
import ai.routin.mytoken.domain.model.UsageMetric
import ai.routin.mytoken.domain.model.UsageMetricHealthState
import ai.routin.mytoken.domain.model.UsageMetricPresentation
import ai.routin.mytoken.domain.model.UsageMetricSemantic
import ai.routin.mytoken.domain.model.UsageMetricUnit
import ai.routin.mytoken.domain.model.UsageSnapshot
import ai.routin.mytoken.domain.usage.UsageProvider
import ai.routin.mytoken.domain.usage.UsageProviderException
import ai.routin.mytoken.provider.health.UsageMetricHealthEvaluator
import ai.routin.mytoken.provider.http.HttpTransport
import ai.routin.mytoken.provider.http.ProviderHttpRequest
import ai.routin.mytoken.provider.http.ProviderHttpResponse
import ai.routin.mytoken.provider.json.ProviderJson
import ai.routin.mytoken.provider.json.arrayOrNull
import ai.routin.mytoken.provider.json.asObjectOrNull
import ai.routin.mytoken.provider.json.boolOrNull
import ai.routin.mytoken.provider.json.decimalOrNull
import ai.routin.mytoken.provider.json.intOrNull
import ai.routin.mytoken.provider.json.longOrNull
import ai.routin.mytoken.provider.json.objOrNull
import ai.routin.mytoken.provider.json.stringOrNull
import java.math.BigDecimal
import java.net.URI
import java.net.URLEncoder
import java.time.Clock
import java.time.Instant
import kotlinx.coroutines.async
import kotlinx.coroutines.coroutineScope
import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import kotlinx.serialization.json.JsonObject

/**
 * Command Code adapter using the same account endpoints as the official CLI:
 * whoami, billing credits, subscriptions, usage summary, and the public model list.
 */
class CommandCodeUsageProvider(
    private val transport: HttpTransport,
    private val clock: Clock = Clock.systemUTC(),
    private val baseURL: URI = URI.create("https://api.commandcode.ai"),
) : UsageProvider {

    override val providerId: ProviderId = ProviderId.CommandCode

    private val metadataCache = mutableMapOf<String, CachedMetadata>()
    private val cacheMutex = Mutex()

    override suspend fun fetchUsage(
        credential: Credential,
        secret: CredentialSecret,
    ): Result<UsageSnapshot> {
        if (credential.providerId != ProviderId.CommandCode ||
            credential.credentialKind != CredentialKind.BearerApiKey
        ) {
            return Result.failure(UsageProviderException.InvalidCredential())
        }
        val token = (secret as? CredentialSecret.BearerToken)?.token.orEmpty().trim()
        if (token.isEmpty()) return Result.failure(UsageProviderException.InvalidCredential())

        return runCatching { fetchSnapshot(credential, token) }.recoverCatching { error ->
            if (error is UsageProviderException) throw error
            throw UsageProviderException.InvalidResponse()
        }
    }

    private suspend fun fetchSnapshot(
        credential: Credential,
        token: String,
    ): UsageSnapshot {
        val now = clock.instant()
        val cacheKey = "${credential.id}:${token.hashCode()}"
        val cached = cacheMutex.withLock { metadataCache[cacheKey] ?: CachedMetadata() }

        return coroutineScope {

        val whoamiRequest = async {
            requestObject("/alpha/whoami", token, mapOf("limits" to "1"))
        }
        val creditsRequest = async {
            requestResult("/alpha/billing/credits", token, query = mapOf("orgId" to cached.orgId))
        }
        val subscriptionsRequest = async {
            requestResult("/alpha/billing/subscriptions", token, query = mapOf("orgId" to cached.orgId))
        }
        val modelsRequest = async {
            fetchAllowedModels(cacheKey, token, now)
        }
        val preloadedSummaryRequest = cached.periodStart?.let { periodStart ->
            async {
                requestResult(
                    "/alpha/usage/summary",
                    token,
                    query = mapOf(
                        "orgId" to cached.orgId,
                        "since" to periodStart,
                    ),
                )
            }
        }

        val whoami = whoamiRequest.await()
        val actualOrgID = whoami.objOrNull("org")?.stringOrNull("id")?.takeIf { it.isNotBlank() }
        val creditsPair = if (actualOrgID == cached.orgId) {
            creditsRequest.await().getOrThrow() to subscriptionsRequest.await().getOrThrow()
        } else {
            creditsRequest.cancel()
            subscriptionsRequest.cancel()
            requestObject("/alpha/billing/credits", token, query = mapOf("orgId" to actualOrgID)) to
                requestObject("/alpha/billing/subscriptions", token, query = mapOf("orgId" to actualOrgID))
        }
        val creditsRoot = creditsPair.first
        val creditsData = creditsRoot.objOrNull("credits")
            ?: throw UsageProviderException.ProviderMessage("Command Code：额度信息返回失败")
        val subscription = creditsPair.second.objOrNull("data")
        val periodStart = subscription?.stringOrNull("currentPeriodStart")
        val summary = if (
            preloadedSummaryRequest != null &&
            actualOrgID == cached.orgId &&
            periodStart == cached.periodStart
        ) {
            preloadedSummaryRequest.await().getOrThrow()
        } else {
            preloadedSummaryRequest?.cancel()
            requestObject(
                "/alpha/usage/summary",
                token,
                query = mapOf(
                    "orgId" to actualOrgID,
                    "since" to periodStart,
                ),
            )
        }
        val allowedModels = modelsRequest.await()

        cacheMutex.withLock {
            val current = metadataCache[cacheKey] ?: CachedMetadata()
            metadataCache[cacheKey] = current.copy(orgId = actualOrgID, periodStart = periodStart)
        }

        val monthlyRemaining = creditsData.decimalOrNull("monthlyCredits")?.max(BigDecimal.ZERO) ?: BigDecimal.ZERO
        val purchasedRemaining = creditsData.decimalOrNull("purchasedCredits")?.max(BigDecimal.ZERO) ?: BigDecimal.ZERO
        val freeRemaining = creditsData.decimalOrNull("freeCredits")?.max(BigDecimal.ZERO) ?: BigDecimal.ZERO
        val totalRemaining = monthlyRemaining + purchasedRemaining + freeRemaining
        val totalSpent = summary.decimalOrNull("totalCost")?.max(BigDecimal.ZERO) ?: BigDecimal.ZERO
        val planID = subscription?.stringOrNull("planId")
        val plan = PlanInfo.from(planID)
        val status = subscription?.stringOrNull("status")
        val active = status == "active" || status == "trialing"
        val totalPool = if (active && plan.monthlyCredits != null) {
            plan.monthlyCredits.max(monthlyRemaining) + purchasedRemaining + freeRemaining
        } else {
            totalSpent + totalRemaining
        }
        val totalUsed = (totalPool - totalRemaining).max(BigDecimal.ZERO)
        val health = UsageMetricHealthEvaluator.balanceState(
            balance = totalRemaining,
            warningThreshold = null,
            isAvailable = true,
        )

        val metrics = buildList {
            add(
                UsageMetric(
                    id = "credit-progress",
                    label = "月",
                    used = totalUsed,
                    limit = totalPool,
                    remaining = totalRemaining,
                    unit = UsageMetricUnit.Currency,
                    presentation = UsageMetricPresentation.Progress,
                    semantic = UsageMetricSemantic.UsedQuota,
                    currencyCode = "$",
                    healthState = health,
                )
            )
            add(valueMetric("credit-balance", "剩余额度", totalRemaining, health))
            add(valueMetric("monthly-remaining", "月度剩余", monthlyRemaining, health))
            add(valueMetric("purchased-remaining", "购买剩余", purchasedRemaining, health))
            add(valueMetric("free-remaining", "赠送剩余", freeRemaining, health))
            add(valueMetric("period-spent", "本周期消费", totalSpent, UsageMetricHealthState.Normal))
            add(
                valueMetric(
                    id = "request-count",
                    label = "累计请求",
                    value = (summary.intOrNull("totalCount") ?: 0).coerceAtLeast(0).toBigDecimal(),
                    health = UsageMetricHealthState.Normal,
                )
            )
            addAll(windowMetrics(creditsRoot.objOrNull("windowLimits")))
        }

        UsageSnapshot(
            credentialId = credential.id,
            fetchedAt = now,
            planName = plan.name,
            subscriptionStartAt = periodStart?.let(Instant::parse),
            subscriptionEndAt = subscription?.stringOrNull("currentPeriodEnd")?.let(Instant::parse),
            statusText = statusText(status),
            usageKind = "periodic",
            billingMode = status,
            allowedModels = allowedModels,
            metrics = metrics,
        )
        }
    }

    private fun windowMetrics(limits: JsonObject?): List<UsageMetric> {
        if (limits?.boolOrNull("limited") != true) return emptyList()
        return listOfNotNull(
            limits.objOrNull("fiveHour")?.let { usageWindow("five-hour", "5 小时", it) },
            limits.objOrNull("weekly")?.let { usageWindow("weekly", "周", it) },
        )
    }

    private fun usageWindow(id: String, label: String, data: JsonObject): UsageMetric {
        val used = (data.decimalOrNull("used") ?: BigDecimal.ZERO).max(BigDecimal.ZERO)
        val cap = (data.decimalOrNull("cap") ?: BigDecimal.ZERO).max(BigDecimal.ZERO)
        return UsageMetric(
            id = id,
            label = label,
            used = used,
            limit = cap,
            remaining = (cap - used).max(BigDecimal.ZERO),
            unit = UsageMetricUnit.Currency,
            windowEnd = data.longOrNull("resetAt")?.let(Instant::ofEpochMilli),
            presentation = UsageMetricPresentation.Progress,
            semantic = UsageMetricSemantic.UsedQuota,
            currencyCode = "$",
            healthState = UsageMetricHealthState.Normal,
        )
    }

    private fun valueMetric(
        id: String,
        label: String,
        value: BigDecimal,
        health: UsageMetricHealthState,
    ) = UsageMetric(
        id = id,
        label = label,
        value = value,
        unit = UsageMetricUnit.Currency,
        presentation = UsageMetricPresentation.Value,
        semantic = UsageMetricSemantic.Value,
        currencyCode = "$",
        healthState = health,
    )

    private suspend fun requestResult(
        path: String,
        token: String,
        query: Map<String, String?> = emptyMap(),
    ): Result<JsonObject> = try {
        Result.success(requestObject(path, token, query))
    } catch (error: kotlinx.coroutines.CancellationException) {
        throw error
    } catch (error: Throwable) {
        Result.failure(error)
    }

    private suspend fun fetchAllowedModels(
        cacheKey: String,
        token: String,
        now: Instant,
    ): List<String> {
        val cached = cacheMutex.withLock { metadataCache[cacheKey] ?: CachedMetadata() }
        val fetchedAt = cached.modelsFetchedAt
        if (fetchedAt != null && now.epochSecond - fetchedAt.epochSecond < MODELS_CACHE_SECONDS) {
            return cached.models
        }

        val models = runCatching {
            requestObject("/provider/v1/models", token).arrayOrNull("data")
                .orEmpty()
                .mapNotNull { item ->
                    item.asObjectOrNull()?.stringOrNull("id")?.takeIf(String::isNotBlank)
                }
        }.getOrElse { error ->
            if (error is kotlinx.coroutines.CancellationException) throw error
            return cached.models
        }
        cacheMutex.withLock {
            val current = metadataCache[cacheKey] ?: CachedMetadata()
            metadataCache[cacheKey] = current.copy(models = models, modelsFetchedAt = now)
        }
        return models
    }

    private suspend fun requestObject(
        path: String,
        token: String,
        query: Map<String, String?> = emptyMap(),
    ): JsonObject {
        val queryString = query
            .filterValues { it != null }
            .entries
            .joinToString("&") { (key, value) ->
                "${encode(key)}=${encode(requireNotNull(value))}"
            }
        val url = buildString {
            append(baseURL.toString().trimEnd('/'))
            append(path)
            if (queryString.isNotEmpty()) append("?").append(queryString)
        }
        val response = try {
            transport.execute(
                ProviderHttpRequest(
                    method = "GET",
                    url = url,
                    headers = mapOf(
                        "Authorization" to "Bearer $token",
                        "Accept" to "application/json",
                    ),
                )
            )
        } catch (error: Exception) {
            if (error is kotlinx.coroutines.CancellationException) throw error
            throw UsageProviderException.Transport()
        }
        when (response.statusCode) {
            401, 403 -> throw UsageProviderException.Unauthorized()
            429 -> throw UsageProviderException.RateLimited()
            in 500..599 -> throw UsageProviderException.ProviderUnavailable()
            !in 200..299 -> throw UsageProviderException.InvalidResponse()
        }
        val root = ProviderJson.parse(response.body.decodeToString()).asObjectOrNull()
            ?: throw UsageProviderException.InvalidResponse()
        if (root.boolOrNull("success") == false) {
            throw UsageProviderException.ProviderMessage(
                "Command Code：${root.objOrNull("error")?.stringOrNull("message") ?: "请求返回失败"}"
            )
        }
        return root
    }

    private fun encode(value: String): String = URLEncoder.encode(value, Charsets.UTF_8.name())

    private fun statusText(status: String?): String? = when (status) {
        "active" -> "有效"
        "trialing" -> "试用中"
        "past_due" -> "逾期"
        "canceled", "cancelled" -> "已取消"
        else -> status?.takeIf { it.isNotEmpty() }
    }

    private data class CachedMetadata(
        val orgId: String? = null,
        val periodStart: String? = null,
        val models: List<String> = emptyList(),
        val modelsFetchedAt: Instant? = null,
    )

    private companion object {
        const val MODELS_CACHE_SECONDS = 24 * 60 * 60
    }

    private data class PlanInfo(
        val name: String,
        val monthlyCredits: BigDecimal?,
    ) {
        companion object {
            fun from(planID: String?): PlanInfo = when (planID) {
                "individual-go" -> PlanInfo("Go", BigDecimal("10"))
                "individual-goat" -> PlanInfo("GOAT", BigDecimal("70"))
                "individual-pro" -> PlanInfo("Pro", BigDecimal("30"))
                "individual-pro-v1" -> PlanInfo("Pro", BigDecimal("80"))
                "individual-provider" -> PlanInfo("Provider", BigDecimal("15"))
                "individual-max" -> PlanInfo("Max", BigDecimal("150"))
                "individual-ultra" -> PlanInfo("Ultra", BigDecimal("300"))
                "teams-pro" -> PlanInfo("Teams Pro", BigDecimal("40"))
                null -> PlanInfo("", null)
                else -> PlanInfo(planID, null)
            }
        }
    }
}
